tools/wake-format/emitter.cpp

```diff
@@ -2060,13 +2060,25 @@ wcl::doc Emitter::walk_type(ctx_t ctx, CSTElement node) {
   //              .concat(ctx));
 }
 
+wcl::doc Emitter::walk_unary_op(ctx_t ctx, CSTElement node) {
+  MEMO(ctx, node);
+  FMT_ASSERT(node.id() == CST_OP, node, "Expected CST_OP");
+
+  MEMO_RET(
+      fmt()
+          .fmt_if_else(TOKEN_OP_INEQUAL, fmt().walk(WALK_TOKEN).space(), fmt().walk(WALK_TOKEN))
```

Comment:
The condition should check if the child token is a primitive expression or not.

Also this would apply to any unary operator in the same precedence bracket as `!` I think

Reply:
That does seem like the better way of doing it.  `!True` reads well, even if it's completely worthless semantically.

Now that I think more about it, I don't really like that this is making an exception for `!` (and potentially the related class of operators).  We should come up with something that works for unary operators across the board, and "primitive (reusing simple?) expression" seems like a good division.

Reply:
Why do we need to check if the child token is a primitive?

> Also this would apply to any unary operator in the same precedence bracket as ! I think

on it :)

Reply:
> We should add a similar rule for unary post fix operators (e.g. unary left associative operators)
That breaks unary `,` no?

`def x = 5 ,` instead of `def x = 5,`?

Reply:
Yeah, but `def xs = (2, fib 3,)` isn't.  Not quite what the best way to solve that is, actually.

Reply:
also, looking at the quickref, it seems like `,` is the _only_ unary postfix operator

Reply:
So you have two categories of operators

Left associative:
  Dot, Quantifier, MulDiv, AddSub, Test, And, Or, Comma (the docs say that comma is right associative but a single comma is left associative)

Right associative:
  Exponent, InEqual, Dollar, Ascription, Assignment, Comma (but not the unary case)

Their category determines if you render it before or after for a unary operator (you don't have to think about this with binary operators). Both cases need to check if the underlying expression is primitive or not to determine if it can avoid the space.

Reply:
Ah sorry about that last comment. Something strange happened on my end.

Reply:
The other annoyance is DOT expressions

```
!myTuple.myField
```

would be better than 

```
! myTuple.myField
```

We might say that application and DOT are also allowed as well as primitive expressions because there are no operators (and thus no unary operators) that are more tightly binding. This would solve most of the list ending issues but would not solve this case:

```
x, y, 3 + 2,
```

would still add the space if we don't special case `,`

Reply:
Ashley and I spoke offline, turns out the docs are a bit off (which you can figure out if you think of more examples like 5 - 6 or -5). Each category has a separate associativity for unary and binary.